### PR TITLE
Adds c++17 and gnu++17 as possible options to Clang and gcc compilers.

### PR DESCRIPTION
**Pull request by euripedesrocha**
- Adds to get_options on both ClangCPPCompiler and GnuCPPCompiler c++17 and gnu++17.

### CODE REVIEW

**Comment by madebr** [comment]
Let's also add `c++2a` and `gnu++2a`.
See [this article](https://www.phoronix.com/scan.php?page=news_item&px=GCC-Patches-CPP2A-CPP20).

**Comment by jpakkane** [comment]
I'm in the process of updating the CI image with a newer clang. Once that is done this can be rebased so the tests will run against the new image and hopefully pass.

**Comment by jpakkane** [comment]
The CI image has been updated. Please rebase to rerun tests. Thanks.

**Comment by jpakkane** [comment]
Hmmm. LLVM on that image is still not new enough. Please update the unit test (`run_unittests.py` line 1837) so that if the compiler id is clang, the version is less than 5, and the string contains a substring `17` the tests just continues to the next iteration. Thanks.

**Comment by euripedesrocha** [comment]
I'll take a look on this.

**Comment by jpakkane** [comment]
ping

**Comment by euripedesrocha** [comment]
So, I looked into the issue and as you said, clang is 4.xx for artful(If I found the correct Dockerfile at ciimage). A quick fix could be change clang -> clang-5.0 on line 6 of that file. It's installed as clang-5.0, we may need to create a symbolic link to fix this in the image.

Since I took more time to try to fix this(work issues :) ) I believe that the line number you told me to change isn't the same anymore, am I correct?

File: run_unittests.py [hunk 1]
```diff
@@ -1973,6 +1973,8 @@ def _test_stds_impl(self, testdir, compiler, p):
             if v != 'none':
                 cmd_std = " -std={} ".format(v)
                 self.assertIn(cmd_std, cmd)
+                if type(compiler).__name__ == ClangCPPCompiler and '17' in cmd_std:
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__

undefined name 'ClangCPPCompiler'

**Comment by jpakkane** [comment]
Test failure seems relevant:

    ======================================================================
    ERROR: test_compiler_cpp_stds (__main__.LinuxlikeTests)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "run_unittests.py", line 2017, in test_compiler_cpp_stds
        self._test_stds_impl(testdir, cpp, 'cpp')
      File "run_unittests.py", line 1976, in _test_stds_impl
        if type(compiler).__name__ == 'ClangCPPCompiler' and '17' in cmd_std and compiler.version < 5:
    TypeError: '<' not supported between instances of 'str' and 'int'

**Comment by euripedesrocha** [comment]
Yes. I intend to fix this. I choose to wait to have some chat before follow on this(and forgot to write to you):
- Did you saw my suggestion to update the ciimage to have clang-5.0 instead of clang?
- Could you review the solution I implemented for the test? I need to have more knowledge of meson's internals.
- Could you point some docs on how to run the CI tests locally? It's much faster to fix this minor issues.

**Comment by jpakkane** [comment]
> Did you saw my suggestion to update the ciimage to have clang-5.0 instead of clang?

Yes. Sadly we also need to support distros that are so old that they don't have clang 5 available. Thus the unit tests need to pass even if `clang` points to a version older than 5.0. 

> Could you review the solution I implemented for the test

The current approach is fine, it just needs the fix for the comparison, preferably by using `mesonlib.version_compare`.

> Could you point some docs on how to run the CI tests locally?

`./run_tests.py` to run all of them and `./run_unittests.py` to only run unit tests.

**Comment by ignatenkobrain** [comment]
Flake8 detected 9 issues on 578af9932a0aad9770727b3b5e7840a7e2bf4a7a
Visit https://sideci.com/gh/19784232/pull_requests/2489 to review the issues.

File: run_unittests.py [hunk 2]
```diff
@@ -1973,6 +1973,10 @@ def _test_stds_impl(self, testdir, compiler, p):
             if v != 'none':
                 cmd_std = " -std={} ".format(v)
                 self.assertIn(cmd_std, cmd)
+                if (type(compiler).__name__ == 'ClangCPPCompiler' and
+                    '17' in cmd_std and
+                    version_compare(compiler.versioni, '<5.0.0')):
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__

visually indented line with same indent as next logical line

**Comment by jpakkane** [comment]
Merged as #2739 with a slightly different implementation. Thanks.

**Pull request closed without merging**